plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TemporalTimeTravelUtil.java

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake.transactionlog;
+
+import io.airlift.units.DataSize;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.plugin.deltalake.transactionlog.checkpoint.LastCheckpoint;
+import io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail;
+import io.trino.spi.TrinoException;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogParser.readLastCheckpoint;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogJsonEntryPath;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.getEntriesFromJson;
+import static io.trino.plugin.deltalake.transactionlog.checkpoint.TransactionLogTail.loadNewTail;
+import static io.trino.spi.StandardErrorCode.INVALID_ARGUMENTS;
+import static java.lang.String.format;
+
+public final class TemporalTimeTravelUtil
+{
+    private TemporalTimeTravelUtil() {}
+
+    public static long findLatestVersionUsingTemporal(TrinoFileSystem fileSystem, String tableLocation, long epochMillis)
+            throws IOException
+    {
+        Optional<LastCheckpoint> lastCheckpoint = readLastCheckpoint(fileSystem, tableLocation);
+        long entryNumber = lastCheckpoint.map(LastCheckpoint::version).orElse(0L);
+
+        String transactionLogDir = getTransactionLogDir(tableLocation);
+
+        Optional<TransactionLogEntries> checkpointEntries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        if (checkpointEntries.isEmpty()) {
+            throw new MissingTransactionLogException(getTransactionLogJsonEntryPath(transactionLogDir, entryNumber).toString());
+        }
+
+        Optional<CommitInfoEntry> commitInfo = checkpointEntries.get().getEntries(fileSystem)
+                .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                .filter(Objects::nonNull)
+                .findFirst();
+
+        // We can not derive useful info from the last check point
+        if (commitInfo.isEmpty()) {
+            return findLatestVersionFromWholeTransactions(fileSystem, tableLocation, epochMillis);
+        }
+
+        if (commitInfo.get().timestamp() == epochMillis) {
+            return entryNumber;
+        }
+
+        if (commitInfo.get().timestamp() < epochMillis) {
+            long tail = searchTowardsTail(fileSystem, entryNumber + 1, transactionLogDir, epochMillis);
+            if (tail >= 0) {
+                return tail;
+            }
+            return entryNumber;
+        }
+        else {
+            long head = searchTowardsHead(fileSystem, entryNumber - 1, transactionLogDir, epochMillis);
```

Comment:
While searchTowardsTail is likely not that long (there aren't that many entries between the checkpoint and the latest transaction log), searchTowardsHead may deal with a great deal of entries.

Please consider taking inspiration from https://github.com/trinodb/trino/issues/21024#issuecomment-2403696494

Reply:
Added the `TemporalTimeTravelUtil #searchTowardsHeadParallel`, chosen to use parallel search logic is pretty simple that if the end entry number is greater than `MAX_LINEAR_SEARCH_SIZE`.
I am ok and open to make `MAX_LINEAR_SEARCH_SIZE` configurable and default to other values